File: services/episode-service/src/main/java/com/lithoapp/episodeservice/service/EpisodeService.java
package com.lithoapp.episodeservice.service;

import com.lithoapp.episodeservice.dto.request.CreateEpisodeRequest;
import com.lithoapp.episodeservice.dto.request.UpdateEpisodeRequest;
import com.lithoapp.episodeservice.dto.response.EpisodeResponse;
import com.lithoapp.episodeservice.dto.response.EpisodeSummaryResponse;
import com.lithoapp.episodeservice.enums.EpisodeStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface EpisodeService {

    /**
     * Opens a new case folder for a patient. Initial status is always ACTIVE.
     */
    EpisodeResponse createEpisode(CreateEpisodeRequest request);

    /**
     * Returns full episode details by ID.
     */
    EpisodeResponse getEpisodeById(Long id);

    /**
     * Returns a paginated list of episode summaries for a given patient.
     *
     * @param patientId the patient whose episodes are listed
     * @param status    optional filter — null returns all episodes, ACTIVE or CLOSED filters by status
     * @param pageable  pagination and sorting parameters
     */
    Page<EpisodeSummaryResponse> getEpisodesByPatient(Long patientId, EpisodeStatus status, Pageable pageable);

    /**
     * Partially updates the episode case-folder fields (title, notes, recurrence, status).
     * patientId and openedAt are immutable after creation.
     */
    EpisodeResponse updateEpisode(Long id, UpdateEpisodeRequest request);
}

File: frontend/frontLitho/src/features/episodes/services/episode.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import {
  CreateEpisodeRequest,
  EpisodePage,
  EpisodeResponse,
  EpisodeStatus,
  UpdateEpisodeRequest
} from '../models/episode.model';

@Injectable({
  providedIn: 'root'
})
export class EpisodeService {
  private readonly baseUrl = `${environment.apiBaseUrl}/episodes`;

  constructor(private http: HttpClient) {}

  getEpisodesByPatient(
    patientId: number,
    page: number = 0,
    size: number = 20,
    status?: EpisodeStatus
  ): Observable<EpisodePage> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', 'openedAt,desc');

    if (status) {
      params = params.set('status', status);
    }

    return this.http.get<EpisodePage>(`${this.baseUrl}/patient/${patientId}`, { params });
  }

  getEpisodeById(id: number): Observable<EpisodeResponse> {
    return this.http.get<EpisodeResponse>(`${this.baseUrl}/${id}`);
  }

  createEpisode(request: CreateEpisodeRequest): Observable<EpisodeResponse> {
    return this.http.post<EpisodeResponse>(this.baseUrl, request);
  }

  updateEpisode(id: number, request: UpdateEpisodeRequest): Observable<EpisodeResponse> {
    return this.http.put<EpisodeResponse>(`${this.baseUrl}/${id}`, request);
  }
}
